chain/starknet/src/data_source.rs

```diff
@@ -0,0 +1,320 @@
+        }
+        if !event_handlers.is_empty() {
+            kinds.insert(EVENT_HANDLER_KIND);
+        }
+
+        kinds
+    }
+
+    fn match_and_decode(
+        &self,
+        trigger: &StarknetTrigger,
+        block: &Arc<codec::Block>,
+        _logger: &Logger,
+    ) -> Result<Option<TriggerWithHandler<Chain>>, Error> {
+        if self.start_block() > block.number() {
+            return Ok(None);
+        }
+
+        let handler = match trigger {
+            StarknetTrigger::Block(_) => match self.mapping.block_handlers.first() {
+                Some(handler) => handler.handler.clone(),
+                None => return Ok(None),
+            },
+            StarknetTrigger::Event(event) => match self.handler_for_event(event) {
+                Some(handler) => handler.handler,
+                None => return Ok(None),
+            },
+        };
+
+        Ok(Some(TriggerWithHandler::<Chain>::new(
+            trigger.clone(),
+            handler,
+            block.ptr(),
+        )))
+    }
+
+    fn name(&self) -> &str {
+        &self.name
+    }
+
+    fn kind(&self) -> &str {
+        &self.kind
+    }
+
+    fn network(&self) -> Option<&str> {
+        Some(&self.network)
+    }
+
+    fn context(&self) -> Arc<Option<DataSourceContext>> {
+        Arc::new(None)
+    }
+
+    fn creation_block(&self) -> Option<BlockNumber> {
+        None
+    }
+
+    fn is_duplicate_of(&self, other: &Self) -> bool {
+        let DataSource {
+            kind,
+            network,
+            name,
+            source,
+            mapping,
+        } = self;
+
+        kind == &other.kind
+            && network == &other.network
+            && name == &other.name
+            && source == &other.source
+            && mapping.event_handlers == other.mapping.event_handlers
+            && mapping.block_handlers == other.mapping.block_handlers
+    }
+
+    fn as_stored_dynamic_data_source(&self) -> StoredDynamicDataSource {
+        // FIXME (Starknet): Implement me!
+        todo!()
+    }
+
+    fn from_stored_dynamic_data_source(
+        _template: &DataSourceTemplate,
+        _stored: StoredDynamicDataSource,
+    ) -> Result<Self, Error> {
+        // FIXME (Starknet): Implement me correctly
+        todo!()
+    }
+
+    fn validate(&self) -> Vec<Error> {
+        Default::default()
+    }
+
+    fn api_version(&self) -> semver::Version {
+        semver::Version::new(0, 0, 5)
+    }
+
+    fn runtime(&self) -> Option<Arc<Vec<u8>>> {
+        Some(self.mapping.runtime.clone())
+    }
+}
+
+impl DataSource {
+    /// Returns event trigger if an event.key matches the handler.key and optionally
+    /// if event.fromAddr matches the source address. Note this only supports the default
+    /// Starknet behavior of one key per event.
+    fn handler_for_event(&self, event: &StarknetEventTrigger) -> Option<MappingEventHandler> {
+        let event_key = FieldElement::from_byte_slice_be(event.event.keys.first()?).ok()?;
```

Comment:
if the event only matches against 1 key maybe we could either filter the rest out on ingress (when decoding)

Reply:
I agree that this is not optimal but I'm not sure what's the best approach for this is. `Event` is managed by `prost` and generated from `graph-as-to-rust`. We can technically change the `firehose-starknet` protobuf format to only emit 1 element, but that seems to be sub-optimal too, as the network technically allows multiple keys.

An alternative would be to create a new struct that resembles `Event`, but instead of using a `Vec` for `keys`, it uses `Option<>` to capture only the first element if there's any. For all other fields we can just move from the old `Event` struct.

But then we're just changing from checking whether the first element exists, to whether the `Option<>` is `Some`. I'm not sure whether this is an improvement.

Reply:
As far as I can tell the events are not actually passed down to WASM so maybe we could just drop them at the firehose level. Since The AscTransaction type only cares about type and hash then I think ideally we should avoid sending the unnecessary data

Reply:
We don't need to hold the PR for that, just ensure we follow up on this

Reply:
By dropping them at the firsthose level, do you mean that we simply never emit it in `firehose-starknet` in the first place? I always had this impression that a Firehose implementation should be as faithful to the actual network state as possible, due to the quote I read somewhere that "it should be technically feasible to reconstruct an entire archive node from Firehose data along". Dropping them from `firehose-starknet` seems to go against this principle.

Of course, I'm not an expert on this and would make changes as you see fit. Please let me know your final decision on this. Thanks!

Reply:
Also I don't understand the "events are not sent to WASM" part. Do you mean "event keys"? Cuz I'm quite certain event data are indeed passed?

Reply:
> We don't need to hold the PR for that, just ensure we follow up on it

Just saw this. Got it. But if we need to change the Firehose representation now would be the best time lol. Once people start using it that becomes a breaking change.

Reply:
You are correct, the trigger is indeed passed fully, I was looking for an implement `ToAsc...` for event and didn't see one. Since it's passed to WASM then I think it's fine to keep all the keys, you may match on one but want to use the rest in mappings so let's leave it as is.

Reply:
Sure. Thanks for confirming!